### PR TITLE
fix(ktabs): allow tab anchors to be links [KHCP-13866]

### PR DESCRIPTION
**Pull request by portikM**
# Summary

Addresses: https://konghq.atlassian.net/browse/KHCP-13866

Add `to` property in `Tab` interface to support rendering tabs as links

### CODE REVIEW

File: src/components/KTabs/KTabs.vue [hunk 1]
```diff
@@ -75,11 +75,6 @@ const props = defineProps({
     type: Boolean,
     default: false,
   },
-  anchorTabindex: {
-    type: Number,
-    default: 0,
-    validator: (val: number): boolean => val >= -1 && val <= 32767,
-  },
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:82]: issue: isn't this a breaking change?

File: src/components/KTabs/KTabs.vue [hunk 2]
```diff
@@ -11,21 +11,21 @@
...
           :aria-controls="hidePanels ? undefined : `panel-${tab.hash}`"
           :aria-selected="hidePanels ? undefined : (activeTab === tab.hash ? 'true' : 'false')"
           class="tab-link"
-          :class="{ 'has-panels': !hidePanels, disabled: tab.disabled }"
+          :class="{ disabled: tab.disabled }"
           role="tab"
-          :tabindex="getAnchorTabindex(tab)"
-          @click.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
-          @keydown.enter.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
-          @keydown.space.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:23]: question: why are we removing the keyboard events?

File: src/components/KTabs/KTabs.vue [hunk 3]
```diff
@@ -11,21 +11,21 @@
...
+          :tabindex="!tab.disabled ? 0 : -1"
+          v-bind="tabComponent(tab).attributes"
+          @click="!tab.disabled ? handleTabChange(tab.hash) : undefined"
         >
           <slot :name="`${getTabSlotName(tab.hash)}-anchor`">
-            <span>{{ tab.title }}</span>
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:26]: question: why is this span tag being removed? I'm not sure I understand changing the structure.

File: src/components/KTabs/KTabs.vue [hunk 4]
```diff
@@ -97,12 +92,16 @@ const handleTabChange = (tab: string): void => {
...
+  if (tab.to) {
+    if (typeof tab.to === 'string') {
+      return { tag: 'a', attributes: { href: tab.disabled ? undefined : tab.to } }
+    } else if (typeof tab.to === 'object') {
+      return { tag: 'router-link', attributes: { to: tab.disabled ? undefined : tab.to } }
+    }
   }
 
-  return typeof props.anchorTabindex === 'number' && props.anchorTabindex >= -1 && props.anchorTabindex <= 32767 ? String(props.anchorTabindex) : '0'
+  return { tag: 'button', attributes: { ...(tab.disabled && { disabled: true }) } }
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:104]: issue: this was previously a `div` and now you're returning a `button`

**Review by adamdehaven — CHANGES_REQUESTED** [review]

File: src/components/KTabs/KTabs.vue (hunk 2 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue:23]: It was a `div` and now it's becoming either a `button` or an `a` - both elements support keyboard events natively

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue:104]: Yes, we discussed this with @Justineo and decided this would not be considered a breaking change

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue:26]: I decided to remove it as it's not really necessary

File: docs/components/tabs.md [hunk 5]
```diff
@@ -176,41 +182,6 @@ const tabChange = (hash: string): void => {
 </script>
 ```
 
-### anchorTabindex
```
[the remark this answers is not in the document]
Comment (adamdehaven) [reply docs/components/tabs.md:179]: It's still a breaking change. It should be deprecated instead so we don't have to do a breaking release.

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:26]: This could have downstream impact that you may not be aware of. Let's leave in place.

File: docs/components/tabs.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (portikM) [reply docs/components/tabs.md:179]: Okay, let me bring it back and deprecate it intead.

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:104]: Let's have that discussion here -- changing the DOM structure technically _is_ a breaking change.

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (Justineo) [reply src/components/KTabs/KTabs.vue:26]: I believe we should discourage consumer applications from relying on the DOM structure of our core components, as this makes any upstream changes highly susceptible to breaking. It feels rather unreasonable if we can’t even implement such internal optimizations. 🫠

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (Justineo) [reply src/components/KTabs/KTabs.vue:104]: That depends on the public API we are providing. I believe this should be considered an implementation detail.

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:26]: I'm more advocating for not changing the DOM structure if it isn't absolutely necessary. In this particular case, the "old" functionality (i.e. non-link tabs) doesn't require the change in this PR from `div` to `button`

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (Justineo) [reply src/components/KTabs/KTabs.vue:104]: I think the key lies in how we manage user expectations. We should convey to users as clearly as possible that relying on the DOM structure is unsafe, serves as an escape hatch, and carries the risk of breaking in the future.

In my previous work, as the provider of the component library, we would even periodically scan the code of consumer applications to identify potentially risky usage patterns (such as deep selectors) and provide reports to the users, further clarifying the associated risks.

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:104]: If you look at our own internal implementations of Kongponents, using `KButton` as an example, there are hundreds of customizations directly tied to the DOM structure.

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:26]: I’m just being realistic 😅 

This will be exponentially more important with the new portal UI. Customers will likely be using CSS/selectors all over the place; it’s inevitable. 

Discouraging non-critical DOM structure changes is imperative for stability. 

> I decided to remove it as it's not really necessary

TBH this is not the right pattern for an OSS component library.

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (Justineo) [reply src/components/KTabs/KTabs.vue:104]: Kongponents might not be perfectly abstracted, but the risks are internal. For Portal, I think setting clear customization boundaries is key—unlimited customization makes future updates a lot harder.

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:104]: > unlimited customization makes future update

I agree; however, I still am against trivial changes that aren't related to a specific feature or critical fix. We shouldn't just be changing things "because we want to"

Comment (Justineo) [reply src/components/KTabs/KTabs.vue:104]: > We shouldn't just be changing things "because we want to"

I agree with this. However, I also recognize that the changes being made here are not arbitrary. The intention is to improve the project.

> Yes, we discussed this with @Justineo and decided this would not be considered a breaking change

As for my earlier suggestion: rendering `<a>` elements for tabs with `to` shouldn’t be considered a breaking change, as `to` wasn’t supported before. However, changing a `<div>` to a `<button>` does align with the best practices outlined in [WAI's examples](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/#examples). How about keeping this change? @adamdehaven

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:104]: The change to `button` is arbitrary if the proper `role` (and other) attributes are in place on the existing `div`, but yes, does require adding additional things you "get for free" with a button.

I'm just attempting to stress we need to be very stringent with DOM changes as it could very easily break downstream projects. I'll unblock the change if this is how it's preferred to go forward on this PR, but subjective changes should be pushed back on overall.

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:104]: > subjective changes

To be clear, using a button here is more proper, I'm speaking generally.

File: src/components/KCodeBlock/KCodeBlock.vue [hunk 6]
```diff
@@ -397,7 +397,7 @@ const matchingLineNumbers = ref<number[]>([])
 const currentLineIndex = ref<null | number>(null)
 
 const totalLines = computed((): number[] => Array.from({ length: props.code?.split('\n').length }, (_, index) => index + 1))
-const maxLineNumberWidth = computed((): string => totalLines.value[totalLines.value.length - 1]?.toString().length + 'ch')
+const maxLineNumberWidth = computed((): string => totalLines.value[totalLines.value?.length - 1]?.toString().length + 'ch')
```
Comment (johncowen) [inline src/components/KCodeBlock/KCodeBlock.vue:400]: 👋 Drive-by comment: Is this a bug fix, and if so should it go in a different PR? If not should it be here?

File: src/components/KTabs/KTabs.vue [hunk 7]
```diff
@@ -11,21 +11,21 @@
         class="tab-item"
         :class="{ active: activeTab === tab.hash }"
       >
-        <div
+        <component
```
Comment (johncowen) [inline src/components/KTabs/KTabs.vue:14]: Not sure if this is something you'll want to address in this PR, but if all of the "tabs" are using anchors, it means its essentially a `role="navigation"` (or just a `nav`). It just happens to look like what is generally termed a "tab look and feel". (In regards to line 5 here which I can't make a comment specifically on that line because GH)

File: docs/components/tabs.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (portikM) [reply docs/components/tabs.md:179]: Done.

File: src/components/KTabs/KTabs.vue (hunk 3 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue:26]: Brought the `span` back. I think my prior explanation was poorly phrased, I wanted to remove that `span` to optimize DOM structure in this component but your concern is valid.

File: src/components/KTabs/KTabs.vue (hunk 4 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue]: Changed the element back to div here.

File: src/components/KTabs/KTabs.vue [hunk 8]
```diff
@@ -89,6 +95,8 @@ const emit = defineEmits<{
 
 const activeTab = ref<string>(props.modelValue ? props.modelValue : props.tabs[0]?.hash)
 
+const isNav = computed((): boolean => props.tabs.every(tab => tab.to))
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:98]: question: what happens if some tabs have a `to` prop and others don't?

File: src/components/KTabs/KTabs.vue (hunk 7 above)
Comment (johncowen) [reply src/components/KTabs/KTabs.vue:14]: If you are using the `role="navigation"` you'll probably want to let the user decide what the `aria-label` is.

Question: Would you consider a very simple `KNavigation` component that was just a `<nav><ul><li>` with some similar styling for the tabs (i.e. very minimal JS)?

I would guess a theoretical `KNavigation` would be an very simple component to write as you need about 5% of the JS you have in here. You would also then avoid complicating `KTabs`. I would guess that would also stop @adamdehaven sweating about possible breaking changes 😅 so a positive all round!

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:14]: I think this is getting out of scope for the PR. 

TBH the existing aria-label should be removed as it's generic and doesn't actually describe the content. "Tabs" doesn't label the type of content being displayed.

Comment (portikM) [reply src/components/KTabs/KTabs.vue:14]: I agree that KNavigation is out of scope for this PR.
As for `aria-label`, it is an attribute required for elements with `role="tablist"` and `role="navigation"`. Removing it will bite us back in the upcoming a11y audit for sure. Instead I suggest that we bind `aria-label` (when provided) to this element instead of topmost wrapper `div` and add a paragraph in docs mentioning that it would be a good idea to provide a custom value, or otherwise the default ("Tabs") will be used?

Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:14]: Again, I still think out of scope for this PR. 

Can create a ticket to discuss for a future update

File: src/components/KTabs/KTabs.vue [hunk 9]
```diff
@@ -75,6 +78,9 @@ const props = defineProps({
     type: Boolean,
     default: false,
   },
+  /**
+   * @deprecated
```
[the remark this answers is not in the document]
Comment (portikM) [reply src/components/KTabs/KTabs.vue:82]: I removed this prop from docs (our standard practice for deprecated props is to remove them from Kongponents docs).

[the remark this answers is not in the document]
Comment (adamdehaven) [reply src/components/KTabs/KTabs.vue:82]: > I removed this prop from docs (our standard practice for deprecated props is to remove them from Kongponents docs).

Standard is actually to mark as deprecated 

image

File: src/components/KTabs/KTabs.vue (hunk 8 above)
Comment (portikM) [reply src/components/KTabs/KTabs.vue:98]: Then `role="tablist"` will be used. In other words, means it's a tabs element with some tabs being the links (vs if all tabs are links then it's a navigation element `role="navigation"`).

File: src/components/KTabs/KTabs.vue (hunk 7 above)
Comment (johncowen) [reply src/components/KTabs/KTabs.vue:14]: Ok but if we are using `role=navigation` here make sure you remember to also remove all the other aria attributes and that aren't needed on `role=navigation`. Would also be good to remove or at least make sure that the custom events added to the `li`s here don't break any existing native functionality provided by using anchors.

Whilst discussing this (and this might also be out of scope so let me know if so), I think you might want to remove some of these for when "hide-panels" is true also, as in that case its not really a role="tablist" either.

Let me know the link for that ticket @adamdehaven and I'll catch up with that tomorrow 👍

File: src/components/KTabs/KTabs.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (portikM) [reply src/components/KTabs/KTabs.vue:82]: Well maybe we can't use a word "standard" here since we have different patterns:
* KModalFullscreen is deprecated and we don't have the component docs page
* KTable is deprecated but because it's such a crucial component internally we still have the page up
* `label` prop in KDropdown is deprecated and is not present in docs
* same applies to `isResizable` and `hasError` props in KTextArea
* in fact, same applies to every deprecated prop in any component

So how do we want to go about this one? In my opinion, we don't want to promote usage of it so we should remove it from docs.

File: src/components/KCodeBlock/KCodeBlock.vue (hunk 6 above)
Comment (johncowen) [reply src/components/KCodeBlock/KCodeBlock.vue:400]: Will there be a separate changelog entry for folks doing dependabot PRs and reading changelogs to see what has changed?

Comment (portikM) [reply src/components/KCodeBlock/KCodeBlock.vue:400]: No but a question mark check is hardly a change anyone would be curious about I imagine

Comment (johncowen) [reply src/components/KCodeBlock/KCodeBlock.vue:400]: Oh I thought it was a bug fix?

Comment (portikM) [reply src/components/KCodeBlock/KCodeBlock.vue:400]: Well, DD error fix. Does not affect users.

Comment (johncowen) [reply src/components/KCodeBlock/KCodeBlock.vue:400]: oki doki, feel free to resolve!

File: docs/components/tabs.md [hunk 10]
```diff
@@ -297,6 +268,43 @@ const tabs = ref<Tab[]>([
 </script>
 ```
 
+## Usage
+
+### Tab links
```
Comment (adamdehaven) [inline docs/components/tabs.md:273]: I'm not sure I understand the new "Usage" section here -- can we just document "Tabs as Links" under the Props instead (above the Slots section)

File: docs/components/tabs.md [hunk 11]
```diff
@@ -297,6 +268,43 @@ const tabs = ref<Tab[]>([
...
+
+Passing `to` property for each tab object allows to render tabs as links. If a string is passed, it will be used in `href` attribute in the rendered `a` element. If an object is passed, the tab will be rendered as a `router-link`.
```
Comment (adamdehaven) [inline docs/components/tabs.md:275]: Can you reword to explain this is an example of utilizing the `to` property along with the `hidePanels` prop?

The way it's currently written it sounds like you **must** use as links

File: src/components/KTabs/KTabs.vue [hunk 12]
```diff
@@ -49,7 +52,7 @@
 
 <script lang="ts" setup>
 import type { PropType } from 'vue'
-import { ref, watch } from 'vue'
+import { computed, ref, watch } from 'vue'
```
Comment (adamdehaven) [inline src/components/KTabs/KTabs.vue:55]: remove unused import

File: docs/components/tabs.md (hunk 11 above)
Comment (portikM) [reply docs/components/tabs.md]: Rewritten

File: docs/components/tabs.md (hunk 10 above)
Comment (portikM) [reply docs/components/tabs.md]: https://deploy-preview-2532--kongponents.netlify.app/components/tabs.html#tabs-as-links

**Review by adamdehaven — APPROVED** [review]
